app/src/main/java/com/gruppo3/wetravel/mapManager/DirectionsManager.java

```diff
@@ -0,0 +1,167 @@
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * Singleton class that downloads JSON data from Google Maps Directions Web Service, compute it and shows the corresponding {@link com.google.android.gms.maps.model.Polyline} in a map fragment.
+ *
+ * @author Giovanni Barca
+ */
+public class DirectionsManager {
+    /**
+     * Possible path computing options using Google Maps web service.
+     * <li>
+     *     <ul>DRIVING: Uses roads suitable to car transit;</ul>
+     *     <ul>WALkING: Uses roads and walk paths;</ul>
+     *     <ul>BICYCLING: Uses roads, walk paths and cycle paths;</ul>
+     *     <ul>TRANSIT: Uses mass transit available in a specific zone combined with walking paths.</ul>
+     * </li>
+     */
+    public enum DirectionModes {
+        DRIVING,
+        WALKING,
+        BICYCLING,
+        TRANSIT
+    }
+
+    private static final DirectionsManager instance = new DirectionsManager();
+
+    /**
+     * Gets the instance of this singleton class.
+     * @return Instance of this singleton class
+     */
+    public static DirectionsManager getInstance() {
+        return instance;
+    }
+
+    /**
+     * Computes and shows on the map the requested route from origin to destination with a 12px wide blue line.
+     * @param googleMap Object of type {@link GoogleMap} where the polyline will be displayed. Never null.
+     * @param origin Route's origin coordinates. Never null.
+     * @param dest Route's destination coordinates. Never null.
+     * @param directionMode Specifies in which mode the route has to be calculated (driving, walking, bicycle or transit mode). Never null.
+     */
+    public void computeRoute(@NonNull final GoogleMap googleMap, @NonNull LatLng origin, @NonNull LatLng dest, @NonNull DirectionModes directionMode) {
+        String googleDirectionsApiUrl = buildDirectionsUrl(origin, dest, directionMode); // Builds the Google Maps Directions API request url
+
+        // Downloads the JSON from Google Maps Directions Web Service, parses it and sends to showRoutFromParsedJSON(String) if there were no errors
+        new DownloadTask(new DownloadTask.AsyncResponse() {
+            @Override
+            public void onFinishResult(String result) {
+                // Continue if download successfully terminated
+                if (!result.equals(DownloadTask.ERROR_MESSAGE)) {
+                    new ParserTask(new ParserTask.AsyncResponse() {
+                        @Override
+                        public void onFinishResult(List<List<HashMap<String, String>>> result) {
+                            if (result.size() > 0) // If there were errors, list size is 0
+                                createPolylineFromParsedJSON(googleMap, result, 12, Color.BLUE);
+                        }
+                    }).execute(result);
+                }
+            }
+        }).execute(googleDirectionsApiUrl);
+    }
```

Comment:
I see you made an Overload of the method below... But I'm seeing duplicate code in here.
How about simplify your life, by turning all of this in one easy line?
```suggestion
    public void computeRoute(@NonNull final GoogleMap googleMap, @NonNull LatLng origin, @NonNull LatLng dest, @NonNull DirectionModes directionMode) {
    computeRoute(googleMap, origin, dest, directionMode, DEFAULT_WIDTH, Color.BLUE);
    }
```
Also, that `12` isn't specified... But it looks like a default value, am I right?
Then, how about putting it as a constant field?
`final static int DEFAULT_WIDTH = 12;`